### PR TITLE
Nested styles

### PR DESCRIPTION
**Pull request by grokys**
## What does the pull request do?

Allows styles to be nested, similar to the [CSS Nesting syntax](https://www.w3.org/TR/css-nesting-1/) but using `^` instead of `&` as the nesting selector due to problems with `&`  in XML.

Styles can now have children; these child styles must have a nesting selector `^` in their selector, which will match the parent selector, for example:

```xaml
<Style Selector='Border'>
        <Setter Property='Background' Value='Blue'/>

	<Style.Children>

                <!-- Matches "Border.foo" -->
		<Style Selector='^.foo'>
			<Setter Property='Background' Value='Red'/>
		</Style>

	</Style.Children>
</Style>
```

This feature is going to be required for control theme support. Proposals:

https://github.com/AvaloniaUI/Avalonia/issues/2769
https://github.com/AvaloniaUI/Avalonia/discussions/7378

This PR also updates a couple of control templates to use nested styles, namely `Button` and `CheckBox`; mainly for testing.

## Limitations/Questions

- Nesting selector can only appear at the start of the selector currently - do we need to support it elsewhere for an MVP? Use cases?
- ~~The CSS syntax of `&` isn't very XML-friendly as it needs to be written `&amp;`. Do we want a different syntax to CSS?~~ We've decided on `^` I believe
- Children need to be enclosed in a `<Style.Children>` element in XAML - this is due to a limitation of XAML in that one can only have a single `Content` property (which is `Setters). Is this acceptable or do we want to special case this in the XAML compiler?

## TODO

- [x] Validation that nesting selector appears where it should
- [x] Validation that nesting selector doesn't appear where it shouldn't
- [x] Caching for nested styles
- ~~Reuse of parent style activator subscriptions where possible (for performance)~~ will do this in a separate PR

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0020073-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by timunie** [comment]
> The CSS syntax of & isn't very XML-friendly as it needs to be written &amp;. Do we want a different syntax to CSS?

I personally would prefer a different char. Maybe one can have both, so the developer can decide which one to use.

**Comment by maxkatz6** [comment]
Note that we already use ">" character in our selectors. And it wasn't a problem with VS/Rider XAML editors.
Except third party formatting tools that had to be updated for Avalonia (XAML Styler for instance).

**Comment by robloo** [comment]
Awesome! I did not know CSS already did this...

Two comments:
 1. Can the Style.Children element be optional? The child styles should be treated like content in controls and just automatically pulled into the children collection. This will save space and indentation.
 2. Why is the & required? Isn't it implied by default being a child style? I would just make it optional as well which solves the poor XML syntax issue.

**Comment by robloo** [comment]
I'm also happy to help porting over existing styles. I think it can all be done without breaking changes. So after this PR is merged I'll update all the controls I've added to date and can do more as needed.

**Comment by maxkatz6** [comment]
> Why is the & required? Isn't it implied by default being a child style? I would just make it optional as well which solves the poor XML syntax issue.

First of, yes, CSS has draft in a standard for nested styles. But same syntax js already well known with scss. 
Secondary, there is a distinct difference between ".foo" and "&.foo" nested selectors. 
First one will match children controls, second will match the same control. I.e.:
```xaml
<Style Selector="Button">
    <Style Selector=".foo">
```
Can be translated into ```Button .foo``` (any child control with foo class inside of Button). 
While
```xaml
<Style Selector="Button">
    <Style Selector="&.foo">
```
Into ```Button.foo``` (Button with foo class)

Although, now I think, that we don't actually support ```Button .foo``` because it lacks child type information.

**Comment by robloo** [comment]
> First of, yes, CSS has draft in a standard for nested styles. But the same syntax is already well known with scss.

Well, it's good multiple sources independently converge on the same design/architecture. It generally is a good sign that it is correct. Personally, I'm not well versed in CSS so am still learning.

> Although, now I think, that we don't actually support Button .foo because it lacks child type information.

Yes, it seems Avalonia's use cases are more restrictive and "&" can be implied by default, so optional. However, I was just throwing that out as an idea to solve the syntax issue. If it can be done, great. If not, I'll keep thinking of alternates.

**Comment by robloo** [comment]
Ideas to indicate style inherits from parent (working list):

 1. Use an escaped ampersand `&amp;` as currently supported. This is most correct but most annoying to type and read.
 2. Since Avalonia is more restrictive in use-cases, the ampersand could be optional and implied for child styles when missing. Num.1 is still supported in this case if devs want to be explicit.
 3. Switch to a new symbol as discussed above.
    * `^` Seems like the best symbol to me. It represents an up arrow which strongly idicates look above. It also is similar to UML where "relationship is represented by an arrow with a triangular tip".
    * If you want to make devs hate you, there are alternate Unicode symbols (i.e. `＆`) for ampersand that would probably be allowed in XML: https://en.wikipedia.org/wiki/Ampersand#Encoding_and_display

Actually, all three could be supported at the same time so devs can find what works for them.

**Comment by hez2010** [comment]
How about
```xml
<Style Selector="Button">
    <Style Selector="{StyleSelector .foo, RelativeSource=parent}">
```

**Comment by jp2masa** [comment]
> * `^` Seems like the best symbol to me. It represents an up arrow which strongly idicates look above. It also is similar to UML where "relationship is represented by an arrow with a triangular tip".

Curious, I was about to read your comment and thought "What char would represent this well? ^ for sure, it's an arrow pointing to parent!". So I agree with this, even if `&` is supported, it would be a good option to use `^` and maybe recommend it (unless there's some obvious conflict with CSS or XML, I can't think of any).

**Comment by amwx** [comment]
>`^` Seems like the best symbol to me. It represents an up arrow which strongly idicates look above. It also is similar to UML where "relationship is represented by an arrow with a triangular tip".

I like the caret better personally. 

<br />

>Children need to be enclosed in a `<Style.Children>` element in XAML - this is due to a limitation of XAML in that one can only have a single `Content` property (which is `Setters). Is this acceptable or do we want to special case this in the XAML compiler?

If its not too much work to get the Xaml compiler to work with this, then I'd say do it. Avalonia already simplifies xaml by not requiring every xml tag (e.g.,<Setter.Value>) like WPF/UWP does, so I'm all for not having to type an additional tag. Although, I can also see the lack of separation between setters and nested styles could be confusing for those less familiar with Xaml

<br />

Additional question: Will nested styles work with omitting the parent type for pseudoclasses? Something like:
```Xaml
<Style Selector="Button">
    ...

    <Style.Children>
        <Style Selector="^:pointerover /template/ Border#BorderElement">
        ...
        </Style>
    </Style.Children>
</Style>
```

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0020483-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by grokys** [comment]
Got a few hours to work on this again: yes I agree that `^` would make a better nesting selector. I'll change it to use that.

**Comment by grokys** [comment]
Got a few hours to work on this again: yes I agree that `^` would make a better nesting selector. I'll change it to use that.

**Comment by grokys** [comment]
> Note that we already use ">" character in our selectors. And it wasn't a problem with VS/Rider XAML editors.

Sorry @maxkatz6 I missed this. Unfortunately the `&` character does cause problems and needs to be escaped, even in attribute strings (where `>` doesn't for some reason).

**Comment by grokys** [comment]
> What about adding RelativeSelector on the Style that applies to parent Style?

Could you explain the benefits of this? I'm unable to see why we should have two separate properties. How would this allow you to differentiate between `^.foo` and `^ .foo`? How would this allow potentially putting the nesting selector somewhere else but the beginning in future, e.g. `Button ^.foo`?

> Also +1 for the option to omit the Style.Children tag (similar to what [Content] attribute allows)

The `[Content]` attribute allows you to select a _single_ property as the content property. In this case we'd need two content properties: `Setters` and `Children`, and the XAML compiler would need to choose which of these properties to add the element to based on its types. This isn't part of the XAML spec (though it would be nice to have).

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0020529-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Comment by Sorien** [comment]
when it comes `Style.Children` word children looks wrong they are more `nested style rules` than `children`

**Comment by pr8x** [comment]
@grokys Good points.

>  How would this allow you to differentiate between ^.foo and ^ .foo

Didn't think about that. My hope was to avoid introducing another operator syntax, but it seems like it's the best approach here.

> and the XAML compiler would need to choose which of these properties to add the element to based on the element type

If something like that would be possible it certainly be nice to have. Otherwise it's not that important.

**Comment by grokys** [comment]
> when it comes `Style.Children` word children looks wrong they are more `nested style rules` than `children`

This PR is just using the existing `IStyle.Children` collection that has existed for years. I'd be open to changing the naming but `NestedStyleRules` doesn't work for me ;)

**Comment by robloo** [comment]
I get that the name `Children` seems somewhat strange here. Alternatives I can think of would be `Styles` or `StyleCollection`. However, this is a hierarchy and it's by far the most common in XAML frameworks to use `Child` for one content element and `Children` for many. I think @grokys made the best overall choice here and we'll get used to it quickly.

I also agree requiring `Style.Children` tag is not ideal. If we could get rid of it, it solves the problem of naming for the most part too. However, following the XAML spec is again far more important here and having a special case doesnt make as much sense and will probably cause future issues. So again, I think the design is best now as is and the lesser of two non-ideal options.

**Comment by grokys** [comment]
This PR is ready for review.

Regarding `<Style.Children>` I intend to open a separate issue with a proposal for how to extend XAML to not need it.

**Comment by robloo** [comment]
> Regarding this, WPF does actually have something similar in the form of IAddChild (although it's deprecated and I can't get it to work) so I'll use that as an argument that it's not actually diverging from the XAML spec ;)

I see, didnt know about that at all. If there is a way to do this cleanly, great! It appears this was to mix text with other content at one point too. That got me thinking its similar in concept to some of the things done with TextBlock supporting inlines and formatted text with embedded controls. So I guess there is precident although I don't know much about this.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0020564-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

File: src/Avalonia.Base/Styling/StyleChildren.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
+﻿using System.Collections.ObjectModel;
+using Avalonia.Controls;
+
+namespace Avalonia.Styling
+{
+    internal class StyleChildren : Collection<IStyle>
```
Comment (maxkatz6) [inline src/Avalonia.Base/Styling/StyleChildren.cs:6]: It seems to do the same as AvaloniaList with ForEachItem extension. Basically, the way as have merged dictionaries support - https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Controls/ResourceDictionary.cs#L57.

I suppose custom type makes more sense, if we don't need it to have change notifications

**Review by maxkatz6 — APPROVED** [review]
LGTM 🚀

File: src/Avalonia.Base/Styling/StyleChildren.cs (hunk 1 above)
Comment (maxkatz6) [reply src/Avalonia.Base/Styling/StyleChildren.cs:6]: I suppose custom type makes more sense, if we don't need it to have collection change notifications. Any style insertions/removal will trigger OwnerChanged event.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `0.10.999-cibuild0020608-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

File: src/Avalonia.Base/Styling/StyleChildren.cs (hunk 1 above)
Comment (grokys) [reply src/Avalonia.Base/Styling/StyleChildren.cs:6]: Yeah it was done for performance, basically. Saves an event subscription per collection, and an event invocation per change.

**Pull request merged**